sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java

```diff
@@ -630,6 +634,55 @@ public static void printHelp(PrintStream out, Class<? extends PipelineOptions> i
     }
   }
 
+  public static Map<String, PipelineOptionDescriptor> describe(Set<Class<? extends PipelineOptions>> ifaces) {
+    checkNotNull(ifaces);
+    Map<String, PipelineOptionDescriptor> result = new HashMap<>();
+
+    for (Class<? extends PipelineOptions> iface : ifaces) {
+      CACHE.get().validateWellFormed(iface);
+
+      Set<PipelineOptionSpec> properties = PipelineOptionsReflector.getOptionSpecs(iface);
+
+      RowSortedTable<Class<?>, String, Method> ifacePropGetterTable =
+              TreeBasedTable.create(ClassNameComparator.INSTANCE, Ordering.natural());
+      for (PipelineOptionSpec prop : properties) {
+        ifacePropGetterTable.put(prop.getDefiningInterface(), prop.getName(), prop.getGetterMethod());
+      }
+
+      for (Map.Entry<Class<?>, Map<String, Method>> ifaceToPropertyMap :
+              ifacePropGetterTable.rowMap().entrySet()) {
+        Class<?> currentIface = ifaceToPropertyMap.getKey();
+        Map<String, Method> propertyNamesToGetters = ifaceToPropertyMap.getValue();
+
+        List<String> lists = Lists.newArrayList(propertyNamesToGetters.keySet());
+        lists.sort(String.CASE_INSENSITIVE_ORDER);
+        for (String propertyName : lists) {
+          Method method = propertyNamesToGetters.get(propertyName);
+          // TODO: type representation
+          String printableType = method.getReturnType().getSimpleName();
```

Comment:
@robertwb @mxm any thoughts on how we want to specify the option type in a language agnostic way?

Reply:
Would it work to have a list of pre-defined types alongside with a mapping from Java types to those types?

Reply:
Perhaps we can alternatively just define the accumulation for the option, i.e. the equivalent of `store`, `append`, `store_true`? That's all that seems to be required at this point.

Reply:
Seems like this is the main concern, otherwise String works.

Reply:
Since the job service should not care how the options are constructed, I kept deriving the arg parser action to the client code.